File: src/00016.ml
let rec multiply fact carry new_lst = function
  [] -> if carry = 0 then List.rev new_lst else List.rev (carry::new_lst)
  |h::q -> let temp = h*fact+carry in multiply fact (temp/10) ((temp mod 10)::new_lst) q

let ret = 
  let rec iterate n lst = 
    if n = 1000 then lst else (iterate (n+1) (multiply 2 0 [] lst)) in
  iterate 0 [1] |> List.fold_left (+) 0 

File: src/00016.cpp
#include <iostream>
#include <list>
#include <numeric>

using std::accumulate;
using std::list;
using std::cout;
using std::endl;

void multiply(list<int>& lst, int fact){
    int temp,carry=0;
    for(auto ite=lst.begin();ite!=lst.end();++ite){
        temp = *ite * fact + carry;
        *ite = temp % 10;
        carry = temp/10;
    }
    if(carry) lst.push_back(carry);
}
int main(){
    list<int> lst{1};
    for(auto i=0;i<1000;++i){
        multiply(lst,2);        
    }
    int ret = accumulate(lst.begin(),lst.end(),0);
    cout << ret << endl;
    return 0;
}
